scripts/tests/twister/test_environment.py

```diff
@@ -0,0 +1,579 @@
+# environment.py compares a store_false/true argument to None.
+# This case will never raise an error, until that is fixed.
+#    (
+#        None,
+#        None,
+#        None,
+#        ['--device-flash-with-test'],
+#        '--device-flash-with-test requires --device-testing'
+#    ),
+    (
+        None,
+        None,
+        None,
+        ['--shuffle-tests'],
+        '--shuffle-tests requires --subset'
+    ),
+    (
+        None,
+        None,
+        None,
+        ['--shuffle-tests-seed', '0'],
+        '--shuffle-tests-seed requires --shuffle-tests'
+    ),
+    (
+        None,
+        None,
+        None,
+        ['--coverage-formats', 'html', '--coverage-tool', 'lcov'],
+        '--coverage-formats can only be used when' \
+        ' coverage tool is set to gcovr'
+    ),
+    (
+        None,
+        None,
+        None,
+        ['/dummy/unrecognised/arg'],
+        'Unrecognized arguments found: \'/dummy/unrecognised/arg\'.' \
+        ' Use -- to delineate extra arguments for test binary' \
+        ' or pass -h for help.'
+    ),
+    (
+        None,
+        None,
+        True,
+        [],
+        'By default Twister should work without pytest-twister-harness' \
+        ' plugin being installed, so please, uninstall it by' \
+        ' `pip uninstall pytest-twister-harness` and' \
+        ' `git clean -dxf scripts/pylib/pytest-twister-harness`.'
+    ),
+]
+
+
+@pytest.mark.parametrize(
+    'os_name, which_dict, pytest_installed_var, args, expected_error',
+    TESTDATA_1,
+    ids=[
+        'short build path without ninja',
+        'device-serial-pty on Windows',
+        'west runner without west flash',
+        'west-flash without device-testing',
+        'valgrind without executable',
+        'device serial with multiple platforms',
+#        'device flash timeout without device testing',
+#        'device flash with test without device testing',
+        'shuffle-tests without subset',
+        'shuffle-tests-seed without shuffle-tests',
+        'coverage-formats without gcovr',
+        'unrecognised argument',
+        'pytest-twister-harness installed'
+    ]
+)
+def test_parse_arguments_errors(
+    caplog,
+    os_name,
+    which_dict,
+    pytest_installed_var,
+    args,
+    expected_error
+):
+    def mock_which(name):
+        if name in which_dict['missing']:
+            return False
+        elif name in which_dict['exist']:
+            return which_dict['path'][which_dict['exist']] \
+                if which_dict['path'][which_dict['exist']] \
+                else f'dummy/path/{name}'
+        else:
+            return f'dummy/path/{name}'
+
+    with mock.patch('sys.argv', ['twister'] + args):
+        parser = twisterlib.environment.add_parse_arguments()
+
+    if which_dict:
+        which_dict['path'] = {name: shutil.which(name) \
+            for name in which_dict['exist']}
+        which_mock = mock.Mock(side_effect=mock_which)
+
+    with mock.patch('os.name', os_name) \
```

Comment:
```suggestion
    with mock.patch('os.name', os_name) if os_name is not None else nullcontext(), \
         mock.patch('shutil.which', which_mock) if which_dict else nullcontext(), \
         mock.patch('twisterlib.environment.PYTEST_PLUGIN_INSTALLED', pytest_installed_var) \
            if pytest_installed_var is not None else nullcontext():
```
indentation could be improved to make mocking code more readable.
Could be fine to use [parenthesized-context-manager](https://docs.python.org/3.10/whatsnew/3.10.html#parenthesized-context-managers)
but it is not available in python 3.8

Reply:
I've tried to fit within 80 chars, but that really hurts readability in context managers.
I think that keeping to breaking a single context manager 'line' only on `if` and allowing the text to overrun the 80-char line is a good compromise for now.